kafka/integration_test.go

```diff
@@ -0,0 +1,529 @@
+		if m.TopicPartition.Error != nil {
+			err := m.TopicPartition.Error
+			if testmsg.ExpectedError != "" && err.Error() == testmsg.ExpectedError {
+				fmt.Fprintf(os.Stderr, "OK: Expected error: %s\n", testmsg.ExpectedError)
+			} else {
+				t.Errorf("Encountered unexepected error: %s\n", err)
+			}
+			return
+
+		}
+		// check partition
+		if !((testmsg.Partition == PartitionAny && m.TopicPartition.Partition >= 0) ||
+			(testmsg.Partition >= 0 && m.TopicPartition.Partition == testmsg.Partition)) {
+			t.Errorf("Partition check failed for test message %v. Got back {[%d] [%s] [%s]}\n",
+				testmsg, m.TopicPartition.Partition, m.Key, m.Value)
+			return
+
+		}
+
+		// check Key
+		if testmsg.Key == nil {
+			if !(m.Key == nil || len(m.Key) == 0) {
+				t.Errorf("Key verification failed. Expected %s, got %s\n", testmsg.Key, m.Key)
+				return
+			}
+		} else {
+			if string(m.Key) != string(testmsg.Key) {
+				t.Errorf("Key verification failed. Expected %s, got %s\n", testmsg.Key, m.Key)
+				return
+			}
+		}
+
+		// check Value
+		if testmsg.Value == nil {
+			if !(m.Value == nil || len(m.Value) == 0) {
+				t.Errorf("Value verification failed. Expected %s, got %s\n", testmsg.Value, m.Value)
+				return
+			}
+		} else {
+			if string(m.Value) != string(testmsg.Value) {
+				t.Errorf("Value verification failed. Expected %s, got %s\n", testmsg.Value, m.Value)
+				return
+			}
+		}
+
+		fmt.Printf("%% Message delivered to topic %s [%d] at offset %v: key=(%s), value=(%s)\n",
+			*m.TopicPartition.Topic, m.TopicPartition.Partition, m.TopicPartition.Offset, m.Key, m.Value)
+
+	default:
+		fmt.Printf("Ignored event: %s\n", e)
+	}
+}
+
+// TestConsumerFuncBasedAPI test function-based API, requires broker
+func TestConsumerFuncBasedAPI(t *testing.T) {
+	runConsumerFuncBasedAPI(t, "CommitMessage")
+	runConsumerFuncBasedAPI(t, "CommitOffsets")
+	runProducerChannelBasedAPI(t)
+	runConsumerFuncBasedAPI(t, "Commit")
+}
+
+func runConsumerFuncBasedAPI(t *testing.T, commitMode string) {
+	if !testconfRead() {
+		t.Skipf("Missing testconf.json")
+	}
+	c, err := NewConsumer(&ConfigMap{
+		"bootstrap.servers":    testconf.Brokers,
+		"group.id":             groupId,
+		"enable.auto.commit":   false,
+		"session.timeout.ms":   6000,
+		"default.topic.config": ConfigMap{"auto.offset.reset": "earliest"}})
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to create consumer: %s\n", err)
+		os.Exit(1)
+	}
+
+	t.Logf("Created Consumer %v\n", c)
+
+	err = c.Subscribe(testconf.Topic, nil)
+
+	maxMsgCnt := 3
+	msgcnt := 0
+	run := true
+
+	savedMessages := make([]*Message, maxMsgCnt)
+
+	for run == true {
+		select {
+		default:
+			ev := c.Poll(100)
+			if ev == nil {
+				continue
+			}
+
+			switch e := ev.(type) {
+			case *Message:
+				fmt.Printf("%% Message on %s: key=(%s), value=(%s)\n",
+					e.TopicPartition, e.Key, e.Value)
+				savedMessages[msgcnt] = e
+				msgcnt++
+				if msgcnt >= maxMsgCnt {
+					run = false
+				}
+			case PartitionEOF:
```

Comment:
This isn't safe, it could reach EOF on any partition before having consumed expected number of messages from other partitions.
So simply ignore this event.